### PR TITLE
Initial stab for shadow proxy

### PR DESCRIPTION
**Pull request by cyx**
- [x] Have a config var that we read from which we post all the requests to async
- [x] Have a simple API to manipulate the shadow URLs (zero downtime config restarts)
- [x] Simple benchmarking of before / after performance ?
- [x] Percentage based shadow proxying (or even estimates) ?

### CODE REVIEW

File: drain.go [hunk 1]
```diff
@@ -89,7 +94,8 @@ func (s *server) serveDrain(w http.ResponseWriter, r *http.Request) {
 	batchCounter.Inc(1)
 
 	parseStart := time.Now()
-	lp := lpx.NewReader(bufio.NewReader(r.Body))
+	buf := bytes.Buffer{}
+	lp := lpx.NewReader(bufio.NewReader(io.TeeReader(r.Body, &buf)))
```
Comment (apg) [inline drain.go]: We need to keep all the headers and such, which this doesn't currently do. But, this was more or less the approach I'd like to see.

Guard shadowUrls, provide an HTTP interface to updating them and removing them, which we then can use deploymaster to add / remove stuff. Then, do the env var to catch restarts.

**Comment by cyx** [comment]
Only benchmarking / testing remains.

**Comment by cyx** [comment]
No benchmarking done, but based on the conversation I did add a conservative request timeout of 1s for all shadow requests -- so this in theory should limit the goroutines if we get lagging upstreams.

File: drain.go [hunk 2]
```diff
@@ -44,8 +52,45 @@ var (
...
+
+	for _, u := range urls {
+		u, e := url.Parse(u)
+		if e != nil {
+			err = e
+			continue
+		}
+
+		// We parse the fragment as a percentage of traffic, e.g. #5 == 5%
+		percentage, e := strconv.Atoi(u.Fragment)
```
Comment (apg) [inline drain.go]: Default the percentage, since you aren't bailing?

File: drain.go [hunk 3]
```diff
@@ -44,8 +52,45 @@ var (
...
+		if e != nil {
+			err = fmt.Errorf("!! shadowURL: Unable to parse fragment = %s, assuming 100%", u.Fragment)
+		}
+
+		// Clear the fragement for posting.
+		u.Fragment = ""
+
+		shadowURLs[u.String()] = big.NewInt(int64(percentage))
```
Comment (apg) [inline drain.go]: huh?

File: drain.go [hunk 4]
```diff
@@ -279,5 +325,74 @@ func (s *server) serveDrain(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if err := setShadowURLs(urls); err != nil {
+		log.Println(err)
+	}
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func getShadowURLs() map[string]*big.Int {
```
Comment (apg) [inline drain.go]: I'd expect to be able to introspect URLs so we could more easily tack on another URL.

File: drain.go (hunk 2 above)
Comment (cyx) [inline drain.go]: Yep defaulted to 100.

File: drain.go (hunk 4 above)
Comment (cyx) [inline drain.go]: I guess you mean an endpoint which just shows what's up there?

File: drain.go [hunk 5]
```diff
@@ -279,5 +339,76 @@ func (s *server) serveDrain(w http.ResponseWriter, r *http.Request) {
...
 	parseTimer.UpdateSince(parseStart)
 
+	if len(shadowURLs) > 0 {
+		go amplify(r.Header, buf)
+	}
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (s *server) serveShadowURLs(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "GET" {
```
Comment (apg) [inline drain.go]: Should probably lock here, but this is not really a problem with the scale at which we'll modify and read this value.

**Pull request merged**